Add Storybook

noelledaley opened the pull request:
This PR is moved to https://github.com/hashicorp/vault/pull/6496.

DingoEatingFuzz commented:
This is looking so so good. I have a few suggestions and one question.

**Suggestion 1:** Add some sort of generic styled div thing to make certain components (like buttons, alerts, and most things) have some margin. Having them flush with the sidebar is distracting.
**Suggestion 2:** Some components would benefit from a story where the component is used in context. The "Toggle Button with Content" story does this nicely.
**Suggestion 3:** Automatically loading stories from the filesystem and displaying them alphabetically is most excellent, but if that list gets long enough, it would be helpful to also use sections. Like "Component|Toggle Button" vs. "Forms|Auth Config Form".

**Question:** It's unclear if the notes page is being driven from the markdown file with the same name as the story file or from the jsdoc commentary within a component definition. What do you mean by jsdoc being the source of truth? Are the markdown files unnecessary, are the markdown files necessary but generated?

meirish commented on ui/app/components/alert-banner.js:
```diff
@@ -1,13 +1,27 @@
+ * @property [AlertBanner.type=null] {String} - The banner type. This comes from the message-types helper.
+ * @property message=null {String} - The message to display within the banner.
+ *
+ * @see {@link https://github.com/hashicorp/vault/search?l=Handlebars&q=AlertBanner|Uses of AlertBanner}
```
Can we generate these links and append or prepend them to the MD instead of extracting them here? Seems like the sort of thing that would get missed / left off / easy to get wrong.

noelledaley replied:
that is an excellent idea. madalyn is going to add this to the generator. thank you for the suggestion!

meirish commented on ui/stories/alert-banner.md:
```diff
@@ -0,0 +1,23 @@
+<a name="AlertBanner
```
What're these links up top? Is this something we want?

meirish commented on ui/stories/alert-popup.stories.js:
```diff
@@ -0,0 +1,45 @@
+    () => ({
+      template: hbs`
+        <h5 class="title is-5">Alert Popup</h5>
+        <AlertPopup
```
Can we loop over `messageTypes` here or something similar?

noelledaley replied:
yissss!

meirish commented on ui/stories/i-con.stories.js:
```diff
@@ -0,0 +1,24 @@
+        {{#each types as |type|}}
+          <h5 class="title is-5">{{humanize type}}</h5>
+          <ICon @glyph={{type}} />
+          <br />
```
I think @DingoEatingFuzz's advice of a wrapping div to give some margin should let us get rid of these `<br>`s.

meirish commented on ui/package.json:
```diff
@@ -100,6 +111,7 @@
     "eslint-plugin-prettier": "^3.0.0",
     "flat": "^4.1.0",
     "ivy-codemirror": "2.1.0",
+    "jsdoc-to-markdown": "^4.0.1",
```
If we're not going to use these in tests for now, can we move all of the new deps to `optionalDependencies` so that we don't install all of it when doing tests or release builds?

noelledaley commented:
@DingoEatingFuzz, thank you for your excellent, pointed feedback! i've implemented most of your suggestions, which i'll summarize below:

- **Suggestion 1:** unfortunately it looks like Storybook for ember doesn't yet support custom decorators that would allow us to add a generic styled div. i agree with you that having components flushed with the sidebar is distracting, so for the time being i've added the Storybook Centered addon which centers components vertically and horizontally within the canvas tab. it's not ideal, but i think it's better than what we had before.

- **Suggestion 2:** i separated the HomeLink component into 3 separate stories: 1 to show on its own, another to show with it with the LogoEdition component, and a 3rd to show it rendered in the context of the top navigation (which is the only context in which HomeLink is rendered anyway). were there any other components you felt like needed to be rendered in another context?

- **Suggestion 3:** sections are a great idea. for the time being i've moved the alerts-related components into a section called "Alerts", and i think we'll add more sections as needed. 

regarding your question, the notes pages are populated like so:
1. we use jsdoc2md to generate markdown based on the jsdoc comments inside the component definition.
2. we copy and paste the output of jsdoc2md into the markdown file associated with each story. @madalynrose is working on a bash script that will automatically populate/update the markdown files for us, but for now the process is manual.

thanks again for the review, and let me know if there's anything else we can improve upon or clarify!

noelledaley commented:
I tidied this PR up a bit and moved it to https://github.com/hashicorp/vault/pull/6496.
